src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PETypeParameterSymbol.cs

```diff
@@ -235,6 +236,147 @@ private ImmutableArray<TypeWithAnnotations> GetDeclaredConstraintTypes()
+                if (((_flags & GenericParameterAttributes.NotNullableValueTypeConstraint) != 0))
+                {
+                    return default;
+                }
+            }
+
+            var type = TypeWithAnnotations.Create(typeSymbol);
+            type = NullableTypeDecoder.TransformType(type, constraintHandle, moduleSymbol);
+            type = TupleTypeDecoder.DecodeTupleTypesIfApplicable(type, constraintHandle, moduleSymbol);
+            return type;
+        }
+
+        private static bool? IsNotNullableIfReferenceTypeFromConstraintType(TypeWithAnnotations constraintType, ConsList<PETypeParameterSymbol> inProgress, out bool isNonNullableValueType)
+        {
+            if (!(constraintType.Type is PETypeParameterSymbol typeParameter) ||
+                (object)typeParameter.ContainingSymbol != inProgress.Head.ContainingSymbol ||
+                typeParameter.GetConstraintHandleCollection().Count == 0)
+            {
+                return IsNotNullableIfReferenceTypeFromConstraintType(constraintType, out isNonNullableValueType);
+            }
+
+            bool? isNotNullableIfReferenceType = typeParameter.CalculateIsNotNullableIfReferenceType(inProgress, out isNonNullableValueType);
+
+            if (isNonNullableValueType)
+            {
+                Debug.Assert(isNotNullableIfReferenceType == true);
+                return true;
+            }
+
+            if (constraintType.NullableAnnotation.IsAnnotated() || isNotNullableIfReferenceType == false)
+            {
+                return false;
+            }
+            else if (constraintType.NullableAnnotation.IsOblivious() || isNotNullableIfReferenceType == null)
+            {
+                return null;
+            }
+
+            return true;
+        }
+
+        private bool? CalculateIsNotNullableIfReferenceType(ConsList<PETypeParameterSymbol> inProgress, out bool isNonNullableValueType)
+        {
+            if (inProgress.ContainsReference(this))
+            {
+                isNonNullableValueType = false;
+                return false;
+            }
+
+            if (this.HasValueTypeConstraint)
+            {
+                isNonNullableValueType = true;
+                return true;
+            }
+
+            bool? fromNonTypeConstraints = CalculateIsNotNullableFromNonTypeConstraints();
+
+            ImmutableArray<TypeWithAnnotations> constraintTypes = this.GetDeclaredConstraintTypes(inProgress);
+
+            if (constraintTypes.IsEmpty)
+            {
+                isNonNullableValueType = false;
+                return fromNonTypeConstraints;
+            }
+
+            bool? fromTypes = IsNotNullableIfReferenceTypeFromConstraintTypes(constraintTypes, inProgress, out isNonNullableValueType);
+
+            if (isNonNullableValueType)
+            {
+                Debug.Assert(fromTypes == true);
+                return true;
+            }
+
+            if (fromTypes == true || fromNonTypeConstraints == false)
```

Comment:
>if (fromTypes == true || fromNonTypeConstraints == false) [](start = 12, length = 57)

nit: It feels like we're implementing the "strictest top-level nullability from constraints" here again. Might there be a way to factor that out (share with existing implementation)? 
Maybe this comes back to whether we're dealing with pairs of `bool?`, `NullableAnnotations` or `TypeWithAnnotations`... Aligning on one (possibly `NullableAnnotations`) might make things simpler and less custom.